---
title: "`annotatr`: Making sense of genomic regions"
author: "Raymond G. Cavalcante and Hani Habra"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: true
    toc_depth: 3
    number_section: true
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

Genomic regions resulting from next-generation sequencing experiments and bioinformatics pipelines are made more valuable when annotated to genomic features. A SNP occurring in an exon, or an enhancer, is likely of greater interest than one occurring in an inter-genic region. It may be of interest to find that a particular transcription factor overwhelmingly binds in promoters, while another binds mostly in 3â€™UTRs. Hyper-methylation at promoters containing a CpG island may indicate different regulatory regimes in one condition compared to another.

`annotatr` provides *pre-computed* genomic annotations and a set of functions to read, intersect, summarize, and visualize genomic regions in the context of genomic annotations.

# Installation

`annotatr` package source is available at [http://www.github.com/rcavalcante/annotatr](http://www.github.com/rcavalcante/annotatr). The package can be installed directly from GitHub with the [`devtools`](https://cran.r-project.org/web/packages/devtools/index.html) package.

```{r, eval=FALSE}
devtools::install_github('rcavalcante/annotatr')
```

# Annotations

We downloaded CpG island and UCSC knownGene tracks (for hg19, hg38, mm9, and mm10) from the UCSC Genome Browser. The package source stores these files in `data-raw/`. The package source also contains scripts (in `data-raw/`) used to transform the raw data into `GenomicRanges` objects. The final annotations are bundled with the package and are located in `data/`. Supported annotations are listed with `supported_annotations()` after loading the `annotatr` package.

## CpG Annotations

The base CpG island (CGI) track serves as our CpG island annotations. CpG shores are defined as 2Kb upstream/downstream from the ends of the CpG islands, less the CpG islands. CpG shelves are defined as another 2Kb upstream/downstream of the farthest upstream/downstream limits of the CpG shores, less the CpG islands and CpG shores. The remaining genomic regions make up the inter-CGI annotation.

## UCSC knownGenes

The UCSC knownGenes annotations include 1-5Kb upstream of the TSS, the promoter (<1Kb upstream of the TSS), 5'UTR, exons, introns, CDS, 3'UTR, and 5'UTR exons, 5'UTR introns, 3'UTR exons, and 3'UTR introns. The schematic below gives an idea of how the location coordinates in the knownGene files can be used to determine the annotations.

![Schematic of knownGene annotations.](annotatr_knownGenes.pdf)

# Usage

## Reading A File

`annotatr` expects that the genomic regions of interest are encoded as a BED6 file ([spec](https://genome.ucsc.edu/FAQ/FAQformat.html#format1)) with no header row. However, we allow the user to overload both the `name` and `score` columns. The `name` column can be a non-unique character vector indicating some classification for each region (e.g. hyper-methylated and hypo-methylated). The `score` column can take continuous or discrete numerical values that represent data associated with each region (e.g. the percent methylation).

```{r}
# This file in inst/extdata represents differentially methylated
# regions between two conditions. The name column is the DM
# classification, and the score column is the methylation difference.
dm_file = system.file(
  'extdata',
  'IDH2mut_v_NBM_names_scores_chr9.txt.gz',
  package='annotatr')
regions = read_bed(
  filename = dm_file,
  genome = 'hg19',
  stranded = FALSE,
  use.score = TRUE)
print(regions)
```

## Intersecting Regions with Annotations

Users may select their annotations a la carte by using the accessors listed with `supported_annotations()`, or by shortcuts. The `cpgs` shortcut annotates regions to CpG islands, CpG shores, CpG shelves, and inter-CGI. The `basic_genes` shortcut annotates regions to 1-5Kb, promoters, 5'UTRs, exons, introns, and 3'UTRs. The `detailed_genes` shortcut annotates regions to 1-5Kb, promoters, 5'UTR exons, 5'UTR introns, CDS exons, CDS introns, 3'UTR exons, and 3'UTR introns. The `basic_genes` and `detailed_genes` shortcuts may not be used at the same time.

```{r}
# Describe the annotations
annots = c('cpgs','basic_genes')
# Intersect the regions we read in with the annotations
intersections = intersect_annotations(
  regions = regions,
  annotations = annots,
  genome = 'hg19',
  ignore.strand = FALSE)
# A list of Hits objects is returned, here is the first item.
print(intersections[1])
# We next annotate the intersections
annotated_intersections = annotate_intersections(
  regions = regions,
  intersections = intersections,
  use.score = TRUE)
# A dplyr::tbl_df object is returned
print(annotated_intersections)
```

## Summarizing Over Annotations

The dataset used in the running example actually has a classification (DMup, DMdown, noDM) and a score (methylation difference between two groups) associated with each region. We will employ both summarization functions in preparation for visualization.

```{r}
# Take the mean of the score column across all regions
# occurring in an annotation.
score_summarized_intersections = summarize_score(annotated_intersections)
print(score_summarized_intersections)

# Count the occurrences of classifications in the name
# column across the annotations.
name_summarized_intersections = summarize_name(annotated_intersections)
print(name_summarized_intersections)
```

## Visualizing Over Annotations

`annotatr` has two visualization functions (`visualize_score()` and `visualize_name()`) that are paired with the summarization functions. The visualization functions return an object of type `ggplot` that can be viewed (`print`) and saved (`ggsave`). For maximum flexibility, the user can pass an `annotation_order` parameter into both visualization functions to ensure the desired order, and to visualize a subset of the annotations. A `data_order` can be passed into `visualize_name()` to order the data classifications as desired, and to visualize a subset of the data classifications.

```{r, fig.align='center', fig.cap='Methylation difference distributions across the CpG annotations.', fig.height=6, fig.width=6, fig.show = 'hold'}
# View the score summarized over the CpG annotations
# A subset of cpgs_order could be chosen to display
# the data distribution only in those annotations.
cpgs_order = c(
  'hg19_cpg_islands',
  'hg19_cpg_shores',
  'hg19_cpg_shelves',
  'hg19_cpg_inter')
visualize_score_cpgs = visualize_score(
  summarized_scores = score_summarized_intersections,
  annotation_order = cpgs_order)
print(visualize_score_cpgs)
```

```{r, fig.align='center', fig.cap='Methylation difference distributions across the basic genes annotation.', fig.height=6.5, fig.width=6.5, fig.show='hold'}
# View the score summarized over the knownGene annotations.
# A subset of genes_order could be chosen to display
# the data distribution only in those annotations.
genes_order = c(
  'hg19_knownGenes_1to5kb',
  'hg19_knownGenes_promoters',
  'hg19_knownGenes_5UTRs',
  'hg19_knownGenes_exons',
  'hg19_knownGenes_introns',
  'hg19_knownGenes_3UTRs')
visualize_score_genes = visualize_score(
  summarized_scores = score_summarized_intersections,
  annotation_order = genes_order)
print(visualize_score_genes)
```

```{r, fig.align='center', fig.cap='Differential methylation classification counts by CpG annotations.', fig.height=6, fig.width=6, fig.show='hold'}
# View the classification summarized over the CpG annotations
# A subset of cpgs_order could be chosen to display
# the data distribution only in those annotations.
cpgs_order = c(
  'hg19_cpg_islands',
  'hg19_cpg_shores',
  'hg19_cpg_shelves',
  'hg19_cpg_inter')
# A subset of cpgs_order could be chosen to display
# the data distribution only in those annotations.
data_order = c(
  'DMup',
  'DMdown')
# The fill = FALSE parameter shows the counts
visualize_name_cpgs = visualize_name(
  summarized_names = name_summarized_intersections,
  annotation_order = cpgs_order,
  data_order = data_order,
  fill = FALSE)
print(visualize_name_cpgs)
```

```{r, fig.align='center', fig.cap='Differential methylation classification proportions by CpG annotations.', fig.height=6, fig.width=6, fig.show='hold'}
# View the classification summarized over the CpG annotations
# A subset of cpgs_order could be chosen to display
# the data distribution only in those annotations.
cpgs_order = c(
  'hg19_cpg_islands',
  'hg19_cpg_shores',
  'hg19_cpg_shelves',
  'hg19_cpg_inter')
# A subset of cpgs_order could be chosen to display
# the data distribution only in those annotations.
data_order = c(
  'DMup',
  'DMdown')
# The fill = TRUE parameter shows the proportions
visualize_name_cpgs = visualize_name(
  summarized_names = name_summarized_intersections,
  annotation_order = cpgs_order,
  data_order = data_order,
  fill = TRUE)
print(visualize_name_cpgs)
```

```{r, fig.align='center', fig.cap='Differential methylation classification proportions by basic genes annotation.', fig.height=6, fig.width=6, fig.show='hold'}
# View the score summarized over the knownGene annotations.
# A subset of genes_order could be chosen to display
# the data distribution only in those annotations.
genes_order = c(
  'hg19_knownGenes_1to5kb',
  'hg19_knownGenes_promoters',
  'hg19_knownGenes_5UTRs',
  'hg19_knownGenes_exons',
  'hg19_knownGenes_introns',
  'hg19_knownGenes_3UTRs')
# A subset of cpgs_order could be chosen to display
# the data distribution only in those annotations.
data_order = c(
  'DMup',
  'DMdown',
  'noDM')
visualize_name_genes = visualize_name(
  summarized_names = name_summarized_intersections,
  annotation_order = genes_order,
  data_order = data_order,
  fill = TRUE)
print(visualize_name_genes)
```
